File: Haskell/VTCLibrary/src/VTC/Tree.hs
module VTC.Tree(Tree, size, insert, contains, makeTree) where

data Tree = Leaf | Node Tree Int Tree deriving (Show)

size :: Tree -> Int
size Leaf = 0
size (Node left _ right) = 1 + size left + size right

insert :: Tree -> Int -> Tree
insert Leaf n = Node Leaf n Leaf
insert (Node left value right) n
    | n == value = Node left value right
    | n <  value = Node (insert left n) value right
    | otherwise  = Node left value (insert right n)
    
contains :: Tree -> Int -> Bool
Leaf `contains` _ = False
(Node left value right) `contains` n
    | n == value = True
    | n <  value = contains left n
    | otherwise  = contains right n
    
makeTree :: Tree
makeTree = Leaf

File: Scala/Hello/src/edu/vtc/cis3030/Tree.scala
package edu.vtc.cis3030

sealed abstract class Tree

case object Leaf extends Tree
case class Node(left: Tree, data: Int, right: Tree) extends Tree

/**
 * Companion object contains various tree processing methods. This example shows a tree of integers.
 */
object Tree {

  class EmptyTreeException(message: String) extends Exception(message)


  def size(t: Tree): Int =
    t match {
      case Leaf => 0
      case Node(left, _, right) => size(left) + size(right) + 1
    }


  def insert(t: Tree, value: Int): Tree = 
    t match {
      case Leaf => Node(Leaf, value, Leaf)
      case Node(left, data, right) =>
        if (value == data) Node(left, data, right)
        else if (value < data) Node(insert(left, value), data, right)
        else Node(left, data, insert(right, value))
    }


  def lookup(t: Tree, value: Int): Boolean =
    t match {
      case Leaf => false
      case Node(left, data, right) =>
        if (value == data) true
        else if (value < data) lookup(left, value)
        else lookup(right, value)
    }


  def min(t: Tree): Int = {
    def minHelper(t: Tree, currentMin: Int): Int =
      t match {
        case Leaf => currentMin
        case Node(left, data, _) => minHelper(left, data)
      }

    if (t == Leaf) throw new EmptyTreeException("No minimum element in an empty tree")
    else minHelper(t, Int.MaxValue)
  }


  def max(t: Tree): Int = {
    def maxHelper(t: Tree, currentMax: Int): Int =
      t match {
        case Leaf => currentMax
        case Node(_, data, right) => maxHelper(right, data)
      }

    if (t == Leaf) throw new EmptyTreeException("No maximum element in an empty tree")
    else maxHelper(t, Int.MinValue)
  }

}
